sw/device/lib/dif/dif_uart.c

```diff
@@ -106,7 +107,17 @@ static dif_uart_config_result_t uart_configure(
   }
 
   // Calculation formula: NCO = 2^20 * baud / fclk.
-  uint64_t nco = ((uint64_t)config->baudrate << 20) / config->clk_freq_hz;
+  // count ones of NCO mask
+  uint32_t nco_width = 0;
+  uint32_t max_nco_value = 1;
+
+  do {
+    max_nco_value = max_nco_value << 1;
+    nco_width += 1;
+  } while ((UART_CTRL_NCO_MASK > max_nco_value) & nco_width != 32);
+
+  uint64_t nco =
+      ((uint64_t)config->baudrate << (nco_width + 4)) / config->clk_freq_hz;
```

Comment:
Seeing as we're +4'ing here, it *may* be worth adding a `_Static_assert` that `nco_width` is at most 28.